SYCL/Matrix/joint_matrix_tensorcores.cpp

```diff
@@ -0,0 +1,286 @@
+
+// REQUIRES: cuda
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple -Xsycl-target-backend --cuda-gpu-arch=sm_80 -DSYCL_EXT_ONEAPI_MATRIX_VERSION=4 %s -o %t.out
+// RUN: %t.out
+//
+// This tests the latest unified matrix extension interfaces.
+// Specifying the sm version via the --cuda-gpu-arch flag is necessary
+// for the Nvidia case.  DPC++ JIT compilation is not
+// supported for the Nvidia matrix extension, although some JIT optimizations
+// are performed at the level of the PTX assembly code.
+
+#include <sycl/sycl.hpp>
+
+using namespace sycl;
+using namespace sycl::ext::oneapi::experimental;
+using namespace sycl::ext::oneapi::experimental::matrix;
+constexpr float bf16_eps = 0.00390625;
+
+// Example usage of Nvidia matrix multiply.
+// Optimizations such as memory paddings for avoiding bank conflicts are not
+// included in this test which aids clarity for what is going on. This example
+// forms a "Big matrix" corresponding to a single "TILE" using cuda example
+// terminology.  Multiple TILES can be used to construct yet larger matrices.
+// This example uses row_major a, b, and accumulator matrices.
+
+// M, N, K define the unit sizes of dimensions of the three types (a, b,
+// accumulator) of matrices per subgroup operation:
+// M: number of rows of "C"/"D" (Accumulator) sub-matrices,
+// number of cols of "B" sub-matrix.
+// N: number of cols of "C"/"D" (Accumulator) sub-matrices,
+// number of rows of "A" sub-matrix.
+// K: number of cols of "A"/number of rows of "B" sub-matrices.
+
+constexpr int N_THREADS_PER_MATRIX_OP =
+    32; // the number of threads per MMA subgroup is always 32 for Nvidia.
+
+constexpr int SUB_TILES_M =
+    2; // number of submatrices per row of accumulator ("C", "D") matrices.
+constexpr int SUB_TILES_N =
+    3; // number of submatrices per col of accumulator matrices.
+constexpr int SUB_TILES_K =
+    4; // number of submatrices per col of "A"/per row of "B", matrices.
+
+template <typename T1, typename T2, size_t M, size_t K, size_t N>
+class TypeHelper;
+
+template <typename T1, typename T2, size_t M, size_t K, size_t N>
+using KernelName = class TypeHelper<T1, T2, M, K, N>;
+
+template <size_t Big_N, size_t Big_K, typename T1, typename T2>
+T2 matrix_ref_mn(const int &m, const int &n, T1 *A, T1 *B, T2 *C) {
+  T2 res = C[m * Big_N + n];
+
+  if constexpr (std::is_same<T1, bfloat16>::value) {
+    for (int k = 0; k < Big_K; k++)
+      res += A[m * Big_K + k] * B[k * Big_N + n];
+  } else {
+    for (int k = 0; k < Big_K; k++)
+
+      res +=
+          static_cast<T2>(A[m * Big_K + k]) * static_cast<T2>(B[k * Big_N + n]);
+  }
+
+  return res;
+}
+
+template <typename T1, typename T2, size_t Sub_Tiles_M, size_t Sub_Tiles_K,
+          size_t Sub_Tiles_N, size_t M, size_t K, size_t N,
+          typename T3 = std::remove_const_t<T1>>
+void test(queue &q) {
+
+  constexpr auto Big_M =
+      Sub_Tiles_M *
+      M; // total number of M dimension matrix elements for the "Big matrix".
```

Comment:
I would prefer having the comment before the declaration to avoid the formatting going mad and splitting it up like this.

Reply:
Thanks, that's better.